File: moe/natj/natj/natj-objctests/src/test/java/objc/binding/ClusterClass.java
/*
Copyright 2014-2016 Intel Corporation

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package objc.binding;


import org.moe.natj.general.NatJ;
import org.moe.natj.general.Pointer;
import org.moe.natj.general.ann.Generated;
import org.moe.natj.general.ann.Library;
import org.moe.natj.general.ann.Owned;
import org.moe.natj.general.ann.Runtime;
import org.moe.natj.objc.ObjCRuntime;
import org.moe.natj.objc.ann.ObjCClassBinding;
import org.moe.natj.objc.ann.Selector;
import apple.NSObject;

@Generated
@Runtime(ObjCRuntime.class)
@ObjCClassBinding
@Library("TestClassesObjC")
public class ClusterClass extends NSObject {
    static {
        NatJ.register();
    }

    @Generated
    protected ClusterClass(Pointer peer) {
        super(peer);
    }

    @Generated
    @Owned
    @Selector("alloc")
    public static native ClusterClass alloc();

    @Generated
    @Selector("init")
    public native ClusterClass init();

    @Generated
    @Selector("initWithString:")
    public native ClusterClass initWithString(String string);

    @Generated
    @Selector("string")
    public native String string();
}

File: natj/natj-objctests/src/test/native/ClusterClass.m
/*
Copyright 2014-2016 Intel Corporation

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#import "ClusterClass.h"

@class ClusterClass;

@interface SubClass : ClusterClass {
    NSString* _string;
}
@end

@implementation ClusterClass

- (id)initWithString:(NSString *)string {
    self = [super init];
    if(self) {
        [self release];
        self = [[SubClass alloc] initWithString:string];
    }
    return self;
}

@end

@implementation SubClass
- (id)initWithString:(NSString *)string {
    self = [super init];
    if(self) {
        _string = [string retain];
    }
    return self;
}

- (NSString*)string {
    return [[_string retain] autorelease];
}

- (void)dealloc {
    [_string release];
    [super dealloc];
}
@end
